android/src/main/java/com/swmansion/rnscreens/ScreensModule.kt

```diff
@@ -0,0 +1,96 @@
+import com.swmansion.rnscreens.events.ScreenTransitionProgressEvent
+import java.util.concurrent.atomic.AtomicBoolean
+
+
+@ReactModule(name = ScreensModule.NAME)
+class ScreensModule(private val mReactContext: ReactApplicationContext) : NativeScreensModuleSpec (
+    mReactContext
+) {
+
+    private var topScreenId: Int = -1
+    private val isActiveTransition = AtomicBoolean(false)
+
+    init {
+        System.loadLibrary("rnscreens")
+        val jsContext = reactApplicationContext.javaScriptContextHolder
+        if (jsContext != null) {
+            nativeInstall(jsContext.get())
+        };
+    }
+
+    private external fun nativeInstall(jsiPtr: Long)
+
+    override fun getName(): String = NAME
+
+    @DoNotStrip
+    private fun startTransition(reactTag: Int?): IntArray {
+        UiThreadUtil.assertOnUiThread()
+        if (isActiveTransition.get() || reactTag == null) {
+            return intArrayOf(-1, -1)
+        }
+        topScreenId = -1
+        val result = intArrayOf(-1, -1)
+        val uiManager = UIManagerHelper.getUIManagerForReactTag(mReactContext, reactTag)
+        val stack = uiManager?.resolveView(reactTag)
+        if (stack is ScreenStack) {
+            val fragments = stack.fragments
+            val screensCount = fragments.size
+            if (screensCount > 1) {
+                isActiveTransition.set(true)
+                stack.attachBelowTop()
+                topScreenId = fragments[screensCount - 1].screen.id
+                result[0] = topScreenId
+                result[1] = fragments[screensCount - 2].screen.id
+            }
+        }
+        return result
+    }
+
+    @DoNotStrip
+    private fun updateTransition(progress: Double) {
+        UiThreadUtil.assertOnUiThread()
+        if (topScreenId == -1) {
+            return
+        }
+        val progressFloat = progress.toFloat();
+        val coalescingKey = (if (progressFloat == 0.0f) 1 else if (progressFloat == 1.0f) 2 else 3).toShort()
+        UIManagerHelper
+            .getEventDispatcherForReactTag(mReactContext, topScreenId)
+            ?.dispatchEvent(
+                ScreenTransitionProgressEvent(
+                    UIManagerHelper.getSurfaceId(mReactContext),
+                    topScreenId, progressFloat, true, true, coalescingKey
+                )
+            )
+    }
+
+    @DoNotStrip
+    private fun finishTransition(reactTag: Int?, canceled: Boolean) {
+        UiThreadUtil.assertOnUiThread()
+        if (!isActiveTransition.get() || reactTag == null) {
+            return
+        }
+        val uiManager = UIManagerHelper.getUIManagerForReactTag(mReactContext, reactTag)
+        val stack = uiManager?.resolveView(reactTag)
+        if (stack is ScreenStack) {
+            if (canceled) {
+                stack.detachBelowTop()
+            } else {
+                stack.detachTop()
+            }
+            isActiveTransition.set(false)
+        }
+        topScreenId = -1
+    }
```

Comment:
Yes, you're right. I understand your point. Since this module is currently responsible for one specific logic domain, it makes sense to keep it as it is. If in the future we find a need to add more functionalities that are not strictly related to screen animation, we can consider splitting this module into sub-classes. I discussed it also with Wojtek. Are you ok with it?